### PR TITLE
Fix exception nulls

### PR DESCRIPTION
**Pull request by roberth**
- `currentExceptionTypeName` has crashed in what appears to be `abi::__cxa_current_exception_type()->` ( https://github.com/hercules-ci/hercules-ci-agent/issues/417). This fixes at least the crash part. (It's still not pretty)

- Some other calls have the potential to crash similarly; also fixed.

- Add a few `const` so we don't have to `strdup` for no good reason. Should be uninteresting. (I tried - lazily - to keep working with `char *` but the demangle function did not like that.)

### CODE REVIEW

File: inline-c-cpp/cxx-src/HaskellException.cxx [hunk 1]
```diff
@@ -46,25 +46,41 @@ const char* HaskellException::what() const noexcept {
...
+    - a mangled string if demangling fails,
+    - "<unknown exception>" if exception type info is not available,
+    - "<no exception>" if no current exception is found.
+ */
 #if defined(__GNUC__) || defined(__clang__)
 const char* currentExceptionTypeName()
 {
+  std::type_info *type_info = abi::__cxa_current_exception_type();
+  if (!type_info)
+    return strdup("<no exception>");
```
Comment (bitonic) [inline inline-c-cpp/cxx-src/HaskellException.cxx:61]: This is unrelated to the PR (it was like this to begin with), but could you update the comment in Exception.hs referencing `free`, while it's actually using `unsafePackMallocCString`, and also note here that these allocations are freed there?

**Review by bitonic — APPROVED** [review]

File: inline-c-cpp/cxx-src/HaskellException.cxx (hunk 1 above)
Comment (bitonic) [reply inline-c-cpp/cxx-src/HaskellException.cxx:65]: They do, thank you

**Comment by roberth** [comment]
Ohh we were waiting for the CI configuration to be done. Shouldn't have merged yet.
